src/ifcopenshell-python/docs/ifcopenshell/getting_started.rst

```diff
@@ -0,0 +1,109 @@
+Getting Started with IFC parsing
+================================
+
+The basis of all parsing and getting information from the IFC starts with obtaining a ``IfcParse::IfcFile`` object and validating that it is good for use.
+
+.. code-block:: c++
+
+    std::string input_file_path = "/path/to/my/model.ifc";
+    IfcParse::IfcFile file(input_file_path);
+    if (!file.good()) {
+      std::cerr << "Unable to parse .ifc file" << std::endl;
+      return 1;
+    }
+
+Schema-agnostic parsing of IFCs
+-------------------------------
+
+It is advisable to design your programme in a schema-agnostic fashion to be able to process all schema versions of input IFCs. 
+While different options are presented `here <http://blog.ifcopenshell.org/2019/12/v060.html>`__, the most comfortable approach for modern C++ developers
+would be with the use of templates as shown below.
+
+.. code-block:: c++
+    
+    // necessary includes
+
+    // Include all possible schema types that could be parsed
+    #include "ifcparse/Ifc2x3.h"
+    #include "ifcparse/Ifc4.h"
+    #include "ifcparse/Ifc4x3_add2.h"
+
+    template<typename Schema>
+    void parseIfc(IfcParse::IfcFile &file) {
+        const typename Schema::IfcProduct::list::ptr elements = file.instances_by_type<typename Schema::IfcProduct>();
+        for (typename Schema::IfcProduct::list::it it = elements->begin();
+            it != elements->end(); ++it) {
+            Schema::IfcProduct *ifcProduct = *it;
+            // TODO: Do something with ifcProduct
+        }
+    }
+    //
+    int main(int argc, char* argv[]) {
+        // file of IfcParse::IfcFile previously defined
+        auto schema_version = file.schema()->name();
+        schema_version = schema_version.substr(3);
+        std::transform(schema_version.begin(), schema_version.end(), schema_version.begin(), [](unsigned char c) { return std::tolower(c); });
+        if (schema_version == "4x3_rc2") {
```

Comment:
Do you think we should rewrite this using the Boost preprocessor mechanism (see below). Either way is a bit unfortunate, you either have a a bunch of if statements, or a incomprehensible bunch of preprocessor magic. I do think that the 2nd approach might in the end result in a better developer experience.

I think this would be the time to decide, as we're setting kind of the basis for the C++ docs here.

https://github.com/IfcOpenShell/IfcOpenShell/blob/9485190604bb3fed6f45199684609667bc2ed1df/src/examples/IfcParseExamples.cpp#L217-L227

Reply:
Hmm, I've never used these Boost preprocessor macros before. I see the advantage of macros, but for both cases, one still has to explicitly enumerate all the available schemas either with a `#define SCHEMA_SEQ` or a `else if `. In my production code, I chain `if`, `else if` and end with a `else` to catch the case where the input file does not match any of those schemas. Would that be possible too with the preprocessor macros?

Reply:
Ok, I'm using preprocessor macros now in an `if-else if` fashion with a final `else` to ensure that one can fail explicitly if the schema is not recognized. We still have the enumerated `#define` for the schemas, but at least it's clearer now. I tested this approach on our productive use-case and it works.